### PR TITLE
ENH: preserve RangeIndex in factorize

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -670,6 +671,8 @@ def factorize(
         na_sentinel = -1
         dropna = False
 
+    if isinstance(values, ABCRangeIndex):
+        return values.factorize(sort=sort)
```
Comment (jreback) [inline pandas/core/algorithms.py:675]: elif

**Comment by jreback** [comment]
minor comment, any perf implications?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
> any perf implications?

```
In [2]: ri = pd.Index(range(10**5))

In [3]: %timeit ri.factorize()
2.28 ms ± 120 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)   # <-- master
33.2 µs ± 199 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)  # <-- PR

In [5]: %timeit mi = pd.MultiIndex.from_arrays([ri, ri, ri])
18.2 ms ± 1.26 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)  # <-- master
308 µs ± 5.21 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <-- PR

In [6]: mi = pd.MultiIndex.from_arrays([ri, ri, ri])

In [7]: %timeit mi.get_loc((10, 10, 10))
13.1 µs ± 65.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)  # <-- master
11 µs ± 116 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)  # <-- PR
```

**Comment by jreback** [comment]
> > any perf implications?
> 
> ```
> In [2]: ri = pd.Index(range(10**5))
> 
> In [3]: %timeit ri.factorize()
> 2.28 ms ± 120 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)   # <-- master
> 33.2 µs ± 199 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)  # <-- PR
> 
> In [5]: %timeit mi = pd.MultiIndex.from_arrays([ri, ri, ri])
> 18.2 ms ± 1.26 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)  # <-- master
> 308 µs ± 5.21 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <-- PR
> 
> In [6]: mi = pd.MultiIndex.from_arrays([ri, ri, ri])
> 
> In [7]: %timeit mi.get_loc((10, 10, 10))
> 13.1 µs ± 65.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)  # <-- master
> 11 µs ± 116 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)  # <-- PR
> ```

nice ok, imagine these progate to other things in a + way.

**Review by jreback — APPROVED** [review]

**Comment by jbrockmendel** [comment]
rebased+green

**Pull request merged**